Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift

```diff
@@ -0,0 +1,519 @@
+        } else {
+            channel.pipeline.fireChannelInactive()
+        }
+        XCTAssert(handler.seenHead)
+        XCTAssert(handler.seenBody)
+        XCTAssert(handler.seenEnd)
+
+        XCTAssertNoThrow(try channel.finish())
+    }
+
+    func testMultipleTEWithChunkedLastHasEOFBodyOnResponseWithChannelInactive() throws {
+        // This is not right: RFC 7230 should allow this as chunked, but http_parser instead parses it as
+        // EOF-terminated. We can't easily override that, so we don't.
+        try assertResponseTransferEncodingHasBodyTerminatedByEOF(transferEncodingHeader: "gzip, chunked", eofMechanism: .channelInactive)
+    }
+
+    func testMultipleTEWithChunkedFirstHasEOFBodyOnResponseWithChannelInactive() throws {
+        // Here http_parser is right, and this is EOF terminated.
+        try assertResponseTransferEncodingHasBodyTerminatedByEOF(transferEncodingHeader: "chunked, gzip", eofMechanism: .channelInactive)
+    }
+
+    func testMultipleTEWithChunkedInTheMiddleHasEOFBodyOnResponseWithChannelInactive() throws {
+        // Here http_parser is right, and this is EOF terminated.
+        try assertResponseTransferEncodingHasBodyTerminatedByEOF(transferEncodingHeader: "gzip, chunked, deflate", eofMechanism: .channelInactive)
+    }
+
+    func testMultipleTEWithChunkedLastHasEOFBodyOnResponseWithHalfClosure() throws {
+        // This is not right: RFC 7230 should allow this as chunked, but http_parser instead parses it as
+        // EOF-terminated. We can't easily override that, so we don't.
+        try assertResponseTransferEncodingHasBodyTerminatedByEOF(transferEncodingHeader: "gzip, chunked", eofMechanism: .halfClosure)
+    }
+
+    func testMultipleTEWithChunkedFirstHasEOFBodyOnResponseWithHalfClosure() throws {
+        // Here http_parser is right, and this is EOF terminated.
+        try assertResponseTransferEncodingHasBodyTerminatedByEOF(transferEncodingHeader: "chunked, gzip", eofMechanism: .halfClosure)
+    }
+
+    func testMultipleTEWithChunkedInTheMiddleHasEOFBodyOnResponseWithHalfClosure() throws {
+        // Here http_parser is right, and this is EOF terminated.
+        try assertResponseTransferEncodingHasBodyTerminatedByEOF(transferEncodingHeader: "gzip, chunked, deflate", eofMechanism: .halfClosure)
+    }
+
+    func testRequestWithTEAndContentLengthErrors() throws {
+        XCTAssertNoThrow(try channel.pipeline.add(handler: HTTPRequestDecoder()).wait())
+
+        // Send a GET with the invalid headers.
+        do {
+            try channel.writeInbound(ByteBuffer(string: "POST / HTTP/1.1\r\nTransfer-Encoding: chunked\r\nContent-Length: 4\r\n\r\n"))
+            XCTFail("Did not throw")
+        } catch HTTPParserError.unexpectedContentLength {
+            // ok
+        } catch {
+            XCTFail("Unexpected error: \(error)")
+        }
+
+        // Must spin the loop.
+        (channel.eventLoop as! EmbeddedEventLoop).run()
```

Comment:
Also call `channel.finish()` and assert return value.

Reply:
Nope, we can't do that here because `channel.finish()` will throw if the channel is already closed.